rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+        raise NotImplementedError("Featurizer must have the capacity to "
+                                  "encode trackers to feature vectors")
+
+    def featurize_trackers(self,
+                           trackers,  # type: List[DialogueStateTracker]
+                           domain  # type: Domain
+                           ):
+        # type: (...) -> Tuple[DialogueTrainingData, List[int]]
+        """Create training data"""
+        self.featurize_mechanism.create_helpers(domain)
+
+        (trackers_as_states,
+         trackers_as_actions,
+         metadata) = self.training_states_and_actions(trackers, domain)
+
+        X, true_lengths = self._featurize_states(trackers_as_states, domain)
+        y = self._featurize_labels(trackers_as_actions, domain)
+
+        return (DialogueTrainingData(X, y, metadata),
+                true_lengths)
+
+    def prediction_states(self,
+                          trackers,  # type: List[DialogueStateTracker]
+                          domain  # type: Domain
+                          ):
+        # type: (...) -> List[List[Dict[Text, float]]]
+        raise NotImplementedError("Featurizer must have the capacity to "
+                                  "create feature vector")
+
+    def create_X(self,
+                 trackers,  # type: List[DialogueStateTracker]
+                 domain  # type: Domain
+                 ):
+        # type: (...) -> Tuple[np.ndarray, List[int]]
+        """Create X for prediction"""
+
+        trackers_as_states = self.prediction_states(trackers, domain)
+        X, true_lengths = self._featurize_states(trackers_as_states, domain)
+        return X, true_lengths
+
+    def persist(self, path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        utils.create_dir_for_file(featurizer_file)
+        with io.open(featurizer_file, 'w') as f:
+            f.write(str(jsonpickle.encode(self)))
+
+    @staticmethod
+    def load(path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        if os.path.isfile(featurizer_file):
+            with io.open(featurizer_file, 'r') as f:
+                _json = f.read()
+            return jsonpickle.decode(_json)
+        else:
+            logger.info("Couldn't load featurizer for policy. "
+                        "File '{}' doesn't exist. ".format(featurizer_file))
+            return None
+
+
+class FullDialogueFeaturizer(Featurizer):
+    def __init__(self, featurize_mechanism):
+        # type: (FeaturizeMechanism) -> None
+        super(FullDialogueFeaturizer, self).__init__(featurize_mechanism)
+
+        self.max_len = None
+
+    def _calculate_max_len(self, as_states):
+        self.max_len = 0
+        for states in as_states:
+            self.max_len = max(self.max_len, len(states))
+        logger.info("The longest dialogue has {} actions."
+                    "".format(self.max_len))
+
+    def _pad_states(self, states):
```

Comment:
so padding moved from the beginning to the end? i think we discussed that, just want to make sure

Reply:
yes for Full dialogue featurizer - yes, for Max history it is the old style and is still done by domain (do we want to change it now?)